### PR TITLE
[coap-secure] fix and test to validate max connection attempt limit

### PR DESCRIPTION
**Pull request by abtink**
This commit contains the following
- It fixes `SecureTransport` to ensure we do the max connection attempt check when transitioning from `kStateCloseNotify`.
- It adds new public `otCoapSecure` APIs to start the agent with a given max connection attempts, and to check the agent's state, whether it is connected, connecting or closed.
- It adds CLI commands under `coaps` for the new APIs.
- It adds `test-026-coaps-conn-limit.py` to validate the max connection attempt limit of CoAP secure agent.

### CODE REVIEW

File: include/openthread/coap_secure.h [hunk 1]
```diff
@@ -87,6 +96,26 @@ typedef void (*otHandleCoapSecureClientConnect)(bool aConnected, void *aContext)
  */
 otError otCoapSecureStart(otInstance *aInstance, uint16_t aPort);
 
+/**
+ * Starts the CoAP secure services and sets the maximum number of allowed connection attempts before stopping the
```
Comment (jwhui) [inline include/openthread/coap_secure.h:100]:
```suggestion
 * Starts the CoAP secure service and sets the maximum number of allowed connection attempts before stopping the
```

File: include/openthread/coap_secure.h [hunk 2]
```diff
@@ -87,6 +96,26 @@ typedef void (*otHandleCoapSecureClientConnect)(bool aConnected, void *aContext)
...
+ * agent automatically.
+ *
+ * @param[in]  aInstance      A pointer to an OpenThread instance.
+ * @param[in] aPort           The local UDP port to bind to.
+ * @param[in] aMaxAttempts    Maximum number of allowed connection request attempts. Zero indicates no limit.
+ * @param[in] aCallback       Callback to notify if max number of attempts has reached and agent is stopped.
+ * @param[in] aContext        A pointer to arbitrary context to use with @p aCallback.
```
Comment (jwhui) [inline include/openthread/coap_secure.h:107]:
```suggestion
 * @param[in]  aInstance      A pointer to an OpenThread instance.
 * @param[in]  aPort          The local UDP port to bind to.
 * @param[in]  aMaxAttempts   Maximum number of allowed connection request attempts. Zero indicates no limit.
 * @param[in]  aCallback      Callback to notify if max number of attempts has reached and agent is stopped.
 * @param[in]  aContext       A pointer to arbitrary context to use with @p aCallback.
```

File: src/cli/README_COAPS.md [hunk 3]
```diff
@@ -87,6 +87,9 @@ coaps response sent
 - [delete](#delete-uri-path-type-payload)
 - [disconnect](#disconnect)
 - [get](#get-uri-path-type)
+- [isconnactive](#isconnactive)
+- [isconnected](#isconnected)
+- [isclosed](#isclosed)
```
Comment (jwhui) [inline src/cli/README_COAPS.md:92]:
```suggestion
- [isclosed](#isclosed)
- [isconnactive](#isconnactive)
- [isconnected](#isconnected)
```

File: src/cli/README_COAPS.md [hunk 4]
```diff
@@ -251,6 +256,36 @@ Stops the application coaps service.
...
+
+```bash
+> coaps isconnected
+yes
+Done
+```
+
+### isconnactive
+
+Indicates whether or not the CoAP secure service connection is active (already connected or establishing a connecting).
```
Comment (jwhui) [inline src/cli/README_COAPS.md:271]:
```suggestion
Indicates whether or not the CoAP secure service connection is active (already connected or establishing a connection).
```

File: tests/toranj/cli/test-026-coaps-conn-limit.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+# -----------------------------------------------------------------------------------------------------------------------
+# Test Implementation
+
+r1.cli('coaps psk RIGHT 1234')
+
+# Set max connection attempts to 3
+r1.cli('coaps start 3')
+
+r1_mleid = r1.get_mleid_ip_addr();
```
Comment (jwhui) [inline tests/toranj/cli/test-026-coaps-conn-limit.py:76]:
```suggestion
r1_mleid = r1.get_mleid_ip_addr()
```

Comment (abtink) [reply tests/toranj/cli/test-026-coaps-conn-limit.py]: 😃

**Review by jwhui — APPROVED** [review]

**Pull request merged**